### PR TITLE
[client] Reset config at app shutdown

### PR DESCRIPTION
**Pull request by damb**
In accordance to #11.

### CODE REVIEW

File: libs/seiscomp/client/application.cpp [hunk 1]
```diff
@@ -1046,9 +1046,16 @@ bool Application::reloadBindings() {
 				SEISCOMP_ERROR("No database query object");
 				return false;
 			}
+		} else {
```
Comment (gempa-jabe) [inline libs/seiscomp/client/application.cpp:1049]: Please move `else` to a new line, see above for an example.

File: libs/seiscomp/client/configdb.cpp [hunk 2]
```diff
@@ -262,6 +262,9 @@ void ConfigDB::load(const char *xml) {
 	ar.close();
 }
 
+void ConfigDB::setConfig(DataModel::Config* config) {
```
Comment (gempa-jabe) [inline libs/seiscomp/client/configdb.cpp:265]: Please attach the asterisk to `config`,  `DataModel::Config *config`.

File: libs/seiscomp/client/configdb.h [hunk 3]
```diff
@@ -52,7 +52,8 @@ class SC_SYSTEM_CLIENT_API ConfigDB {
 			const std::set<std::string>& parameterNames = std::set<std::string>());
 
 		void load(const char *xml);
-
+		
+		void setConfig(DataModel::Config* config);
```
Comment (gempa-jabe) [inline libs/seiscomp/client/configdb.h:56]: Please attach the asterisk to config, see other comment.

**Review by gempa-jabe — CHANGES_REQUESTED** [review]

**Comment by gempa-jabe** [comment]
Great, thanks.

**Comment by damb** [comment]
Note that with this PR, modules not explicitly requiring a loading the config module must call `setLoadConfigModuleEnabled(false)`. This is the default, anyway https://github.com/SeisComP/common/blob/df83362fbb6ea29eeef3f9b2a609e6984f017a2f/libs/seiscomp/client/application.cpp#L273

**Comment by gempa-jabe** [comment]
Why must they call it if it is the default? I can't follow your statement right now (it is Friday ;) ). Does this PR brake current behaviour with respect  to initialization?

**Comment by damb** [comment]
Well, I'm not 100% sure, however, a call to e.g. `bindings2cfg --key-dir ./etc/key -o config.xml` previously succeeded without explicitly specifying a config DB. Now, it fails due to https://github.com/SeisComP/common/blob/df83362fbb6ea29eeef3f9b2a609e6984f017a2f/libs/seiscomp/client/application.cpp#L1050-L1053

if not setting the DB.

**Comment by gempa-jabe** [comment]
Well then we have to fix bindings2cfg.

**Comment by gempa-jabe** [comment]
Fixed with c127b41f8e1fc323148b257ff92afe555ae515e5.

**Comment by damb** [comment]
Thanks.

**Comment by gempa-jabe** [comment]
I have to reopen the PR. The fix with `setConfig` causes an application to crash if it is created after another application has been destroyed and both need to read the configuration. Actually we need to reset the instance and not the config object. I will fix that but check with you if you require `ConfigDB::setConfig(...)` or if it was only necessary for this particular issue.

**Comment by gempa-jabe** [comment]
I have added a `Reset` method to all singletons which are called in the respective applications. That fixes a crash and furthermore also cleans up the loaded plugins.

**Comment by damb** [comment]
@gempa-jabe, thanks for rechecking. The implemented solution works for me, too.

Originally, I used `ConfigDB::setConfig(...)` in order to have it aligned with `Inventory::setInventory(...)`. I wasn't aware of the fact that the underlying config object is reused under certain circumstances.

**Pull request merged**